modules/core/04-channel/ante.go

```diff
@@ -0,0 +1,55 @@
+package channel
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/modules/core/04-channel/keeper"
+	"github.com/cosmos/ibc-go/modules/core/04-channel/types"
+)
+
+type ChannelAnteDecorator struct {
+	k keeper.Keeper
+}
+
+func NewChannelAnteDecorator(k keeper.Keeper) ChannelAnteDecorator {
+	return ChannelAnteDecorator{k: k}
+}
+
+func (cad ChannelAnteDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// do not run redundancy check on DeliverTx or simulate
+	if (ctx.IsCheckTx() || ctx.IsReCheckTx()) && !simulate {
+		// keep track of total packet messages and number of redundancies across `RecvPacket`, `AcknowledgePacket`, and `TimeoutPacket/OnClose`
+		msgs := 0
+		redundancies := 0
+		for _, m := range tx.GetMsgs() {
+			if msg, ok := m.(*types.MsgRecvPacket); ok {
```

Comment:
could we turn this into a switch? would reduce some redundancy (msgs += 1) and I think be easier to read